Update ProcessGroup collective C++ APIs to be non-pure virtual functions

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64943



Most ProcessGroup Collective APIs are pure virtual. As a result, c10d extensions
need to override all of them and throw an error if they don't need certain APIs.
This is too verbose for users. This commit changes those collective APIs to
virtual and throws an error by default. Note that ProcessGroup is still an abstract
class as `getBackendName` is a pure virtual function that all subclasses have
to override.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30906866/)!


Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

mrshenli committed:
[WIP] Update ProcessGroup collective C++ APIs to be non-pure virtual functions

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30906866).

mrshenli committed:
Update on "[WIP] Update ProcessGroup collective C++ APIs to be non-pure virtual functions"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30906866).

mrshenli committed:
Update on "[WIP] Update ProcessGroup collective C++ APIs to be non-pure virtual functions"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Update ProcessGroup collective C++ APIs to be non-pure virtual functions"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Update ProcessGroup collective C++ APIs to be non-pure virtual functions"




cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

cbalioglu commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -201,140 +201,231 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
+      const BroadcastOptions& /* opts */ = BroadcastOptions()) {
+    TORCH_CHECK(
+        false,
+        c10::str(
```
Nit: using `fmt` would be syntactically more clean.

mrshenli replied:
Sure, will do this in the next formatting PR (will also sort the APIs).

cbalioglu commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -201,140 +201,231 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
-    return "undefined";
-  }
+  // Subclasses must override this method to return the backend name
+  virtual const std::string getBackendName() const = 0;
```
Probably it is fairly an edge case since I assume we own all `ProcessGroup`s out there, but technically this is a BC-breaking change. A subclass that did not override `getBackendName()` will now fail to compile due to it being an abstract function.

mrshenli replied:
Right. ProcessGroup C++ is currently a private API. I am trying bring it to prototype in the v1.11 release. Will make a few BC-breaking changes during this process.

cbalioglu reviewed (approved):
Left two minor comments, but overall looks good to me.

mrshenli committed:
Update on "Update ProcessGroup collective C++ APIs to be non-pure virtual functions"


Most ProcessGroup Collective APIs are pure virtual. As a result, c10d extensions need to override all of them and throw an error if they don't need certain APIs. This is too verbose for users. This commit changes those collective APIs to virtual and throws an error by default. Note that ProcessGroup is still an abstract class as `getBackendName` is a pure virtual function that all subclasses have to override.

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Update ProcessGroup collective C++ APIs to be non-pure virtual functions"







cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30906866/)!


Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

mrshenli committed:
Update on "Update ProcessGroup collective C++ APIs to be non-pure virtual functions"




Most ProcessGroup Collective APIs are pure virtual. As a result, c10d extensions
need to override all of them and throw an error if they don't need certain APIs.
This is too verbose for users. This commit changes those collective APIs to
virtual and throws an error by default. Note that ProcessGroup is still an abstract
class as `getBackendName` is a pure virtual function that all subclasses have
to override.

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang cbalioglu gcramer23

Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30906866/)!


Differential Revision: [D30906866](https://our.internmc.facebook.com/intern/diff/D30906866)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30906866).
